### ISSUE TITLE
Unable to getFolderByServerRelativeUrl when path contains # symbol (and presumably others)

### ISSUE THREAD

**Issue by sargearmstrong** [vgrem/phpSPO#198]
Dear Vgrem,

Thanks for this library and efforts! When able, can you kindly advise on the following?

I know it is bad practice to use these types of characters in filepaths, but how do we utilize '#' and '%' in RelativeUrls? Perhaps I'm doing something incorrect, but the following fails to retrieve folders which I presume is related to the RFC 3986 behavior listed in the references below.

_$ctx->getWeb()->getFolderByServerRelativeUrl("Shared Documents/Projects/**EntryID#693**")->getFolders();_

I experimented with various changes in functions getFolderByServerRelativeUrl and ResourcePathServiceOperation, but it always results in either no results, or expression not valid and returns a path without the # portion.

_The expression \"Web/getFolderByServerRelativeUrl('Shared Documents/Projects/EntryID\\" is not valid."}}}_

Thank you!

From April 2020 [Microsoft Docs - Support % and # in files and folders](https://docs.microsoft.com/en-us/sharepoint/dev/solution-guidance/supporting-and-in-file-and-folder-with-the-resourcepath-api)
From February 2017 [Office 365 Dev Blog - Upcoming changes to support # and %](https://developer.microsoft.com/en-us/office/blogs/upcoming-changes-to-sharepoint-and-onedrive-for-business-apis-to-support-and-in-file-names/)

**Comment by vgrem** [comment]
Greetings!

regarding 

> I presume is related to the RFC 3986 behavior listed in the references below

indeed, as the same article explains file addressing might become ambiguous and fixing the existing API to support # and % 

> requires significant changes to our APIs that will impact partners that work with URLs and file names in SharePoint.

In this regard the following options could be considered:


### Option 1

As article states 

> As a recap, note that existing string-based SharePoint APIs (such as SPFileCollection.GetByUrl) handle both encoded and decoded URLs by automatically assuming that % and # characters in a path imply that the URL has been encoded. 

meaning  the existing API method  `FileCollection.getByUrl `method could in fact be utilized for addressing a file which name contains % and # characters as demonstrated below:


$folderUrl = "Shared Documents";
$fileUrl = "Guide #123.docx";
file = $ctx->getWeb()->getFolders()->getByUrl($folderUrl)->getFiles()->getByUrl($fileUrl);
$ctx->load($file);
$ctx->executeQuery();
print "File name: '{$file->getProperty("Name")}'\r\n";


### Option 2

In the next version `(2.4.2` ver)  which adds support for `ResourcePath` and  the method `Web.getFileByServerRelativePath($resourcePath)` in particular, file could be addressed like this:

```
$fileUrl = "/Shared Documents/Sample - #2009.rtf";

$credentials = new ClientCredential($settings['ClientId'], $settings['ClientSecret']);
$file = (new ClientContext($settings['Url']))->withCredentials($credentials)
    ->getWeb()->getFileByServerRelativePath(new ResourcePath($fileUrl))->get()->executeQuery();
```


**Related posts** 

- #96 - Can't get subfolders/subfiles for folder with special chars like # (SharePoint)

- [Sharepoint REST API GetFileByServerRelativeUrl when there's # sign in the file name](https://stackoverflow.com/q/47954004/1375553)


> Note: `2.4.2` which is planned to be released later this week

**Comment by sargearmstrong** [comment]
Thank you for the reply, sorry I missed the existing thread! If I may..

-Does either option support listing the contents of a folder whose name contains the special chars? I played with it some but didn't see how to achieve directory listings when the filename was unknown. 
-Just curious, does the current master 2.4.1 now require PHP > 5.4?  Composer threw a warning and installed 2.4 instead (I'm a little behind in migrating).

Thanks so much!

**Comment by vgrem** [comment]
Please find my comments below


> -Does either option support listing the contents of a folder whose name contains the special chars? I played with it some but didn't see how to achieve directory listings when the filename was unknown.

[Version `2.4.2`](https://github.com/vgrem/phpSPO/tree/v2.4.2) has been released and the following operations are available:

- `Web.getFileByServerRelativePath($resourcePath)`

- `Web.getFolderByServerRelativePath($resourcePath)`

> -Just curious, does the current master 2.4.1 now require PHP > 5.4? Composer threw a warning and installed 2.4 instead (I'm a little behind in migrating).  Positive,  `PHP 5 >= 5.5.0` is required due to due to usage certain features such as Generators

**Comment by F4T4LERROR** [comment]
I just wanted to add that when attempting to reference files/folders using the RelativePath method, I tried using vgrem's examples almost verbatim, but got errors on 2.5.2.  I had to:

Exclude the '/' in front of Shared Documents when defining the URL. 
Use the following syntax: $folder = $ctx->getWeb()->getFolderByServerRelativePath(new SPResourcePath($folderURLl))->select(["Exists"])->get()->executeQuery();

Then, I could check if it exists using if($folder->getExists()) as is typical.
